### PR TITLE
allow teleport to start when some etcd nodes are unreachable

### LINKED ISSUE #23516: Teleport not start if one node etcd is down
Teleport not start if one node etcd is down
My config:
`
peers: ["https://etcd-01.local:2379", "https://etcd-02.local:2379", "https://etcd-03.local:2379"]
`
etcd-01 and etcd-03 is up, and other Teleport  instances work with them.
etcd-02 is down.
```
Mar 23 17:39:06 teleport-02 teleport[5408]: ERROR: initialization failed
Mar 23 17:39:06 teleport-02 teleport[5408]: failed to dial endpoint https://etcd-02.local:2379 with maintenance client: context deadline exceeded
Mar 23 17:39:06 teleport-02 systemd[1]: teleport-server.service: Main process exited, code=exited, status=1/FAILURE
Mar 23 17:39:06 teleport-02 systemd[1]: teleport-server.service: Failed with result 'exit-code'.
```
Teleport v12.1.1 git:api/v12.1.1-0-gb9ee483 go1.19.7

I was expecting Teleport to skip the unreachable node

### PR DESCRIPTION
**Pull request by fspmarshall**
Changes the etcd backend to start if _any_ endpoint is reachable rather than only starting if _all_ endpoints are reachable.

Fixes https://github.com/gravitational/teleport/issues/23516

### CODE REVIEW

File: lib/backend/etcdbk/etcd.go [hunk 1]
```diff
@@ -288,20 +289,8 @@ func New(ctx context.Context, params backend.Params, opts ...Option) (*EtcdBacke
...
-			return nil, trace.Wrap(err)
-		}
-
-		ver := semver.New(status.Version)
-		min := semver.New(teleport.MinimumEtcdVersion)
-		if ver.LessThan(*min) {
-			return nil, trace.BadParameter("unsupported version of etcd %v for node %v, must be %v or greater",
-				status.Version, n, teleport.MinimumEtcdVersion)
-		}
+	if err := b.checkVersion(ctx); err != nil {
```
Comment (zmb3) [inline lib/backend/etcdbk/etcd.go:292]: Should we be using `closeCtx` here and above?

File: lib/backend/etcdbk/etcd.go [hunk 2]
```diff
@@ -315,6 +304,49 @@ func New(ctx context.Context, params backend.Params, opts ...Option) (*EtcdBacke
 	return b, nil
 }
 
+func (b *EtcdBackend) checkVersion(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*9)
```
Comment (zmb3) [inline lib/backend/etcdbk/etcd.go:308]: Why 9 seconds?

File: lib/backend/etcdbk/etcd.go [hunk 3]
```diff
@@ -315,6 +304,49 @@ func New(ctx context.Context, params backend.Params, opts ...Option) (*EtcdBacke
...
+			}
+
+			results <- nil
+		}(nn)
+	}
+
+	// wait for results
+	var err error
+	for range b.cfg.Nodes {
+		err = <-results
```
Comment (zmb3) [inline lib/backend/etcdbk/etcd.go:337]: Should we select over both `results` and `ctx.Done()` here?

File: lib/backend/etcdbk/etcd.go [hunk 4]
```diff
@@ -315,6 +304,49 @@ func New(ctx context.Context, params backend.Params, opts ...Option) (*EtcdBacke
...
+		go func(n string) {
+			defer wg.Done()
+			status, err := b.clients.Next().Status(ctx, n)
+			if err != nil {
+				results <- trace.Wrap(err)
+				return
+			}
+
+			ver := semver.New(status.Version)
+			min := semver.New(teleport.MinimumEtcdVersion)
```
Comment (zmb3) [inline lib/backend/etcdbk/etcd.go:323]: `semver.New` will panic if we fail to parse the version. Should we return the error instead?

File: lib/backend/etcdbk/etcd.go (hunk 3 above)
Comment (fspmarshall) [reply lib/backend/etcdbk/etcd.go:347]: this is blocking on an operation that is scoped to `ctx`, so I don't think it adds anything.

**Review by zmb3 — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**